lib/systemd/system/delphix-migration.service

```diff
@@ -0,0 +1,48 @@
+#
+# Copyright 2019 Delphix
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+#
+# We want to execute this service as early as possible since it will be
+# finalizing the network configuration after an OS migration. Since it will
+# also be importing domain0, which is normally imported by the
+# zfs-import-cache service, we add similar dependencies to it
+# (i.e. devices must be ready, so run after systemd-udev-settle.service, and
+# service is required by zfs-import.target). Finally, we must make sure to
+# remove default service dependencies, as those would create a dependency
+# cycle (this service must run before local-fs.target, but default
+# dependencies require local-fs.target).
+#
+[Unit]
+Description=Delphix OS Migration Service
+DefaultDependencies=no
+After=systemd-udev-settle.service
+Before=network-pre.target
+Before=zfs-import.target
+
+[Service]
+Type=oneshot
+ExecStart=/var/lib/delphix-platform/os-migration
+RemainAfterExit=yes
+
+#
+# Disable the service once it has run.
+# Even though the service does nothing when the flag file is not there,
+# disabling it has the advantage of simplifying systemd dependencies.
+#
+ExecStartPost=/bin/systemctl disable delphix-migration.service
```

Comment:
If we're going to go to the trouble of this, why not use `ConditionFileExists=/var/delphix/migration/perform-migration`? You'd still be able to remove the file and rerun the service this way.

I'm fine with this, I just think it's a more awkward way to accomplish the same thing as using `ConditionFileExists`.

Also, do you want to also "stop" the service, in addition to disabling it, since you're using `RemainAfterExit`? I think the service might appear to be still running, even after this is called. I'm not 100% sure though.

Reply:
The main idea of running the disable is that it removes the service from the systemd dependency graph, bringing its impact down to 0. I haven't seen other ways to achieve this.

Reply:
`ConditionFileExists` doesn't accomplish the same thing? I'm more curious, than trying to state we should use this. I can investigate that later, I guess.

I still wonder if we should stop the service. I also wonder if we should mask the service, instead of disable, but it might not matter either way.

I'm OK with this.

Reply:
In the man files for `systemd.unit` in the `ConditionXXX` section (including `ConditionPathExists` and `ConditionFirstBoot`), it states: 
> Before starting a unit, verify that the specified condition is true. If it is not true, the starting of the unit will be (mostly silently) skipped, however all ordering dependencies of it are still respected.

As for stopping the service I don't think it really matters.